go/pkg/chunker/chunker.go

```diff
@@ -149,47 +156,54 @@ func (c *Chunker) Next() (*Chunk, error) {
 	if !c.HasNext() {
 		return nil, ErrEOF
 	}
-	c.initialized = true
 	if c.digest.Size == 0 {
+		c.reachedEOF = true
 		return &Chunk{}, nil
 	}
-	// Empty contents means we don't have the full data cache, and must read it from a file.
-	if c.contents == nil && c.reader == nil { // Need to initialize the reader.
-		f, err := os.Open(c.path)
-		if err != nil {
-			return nil, err
-		}
-		c.reader = bufio.NewReaderSize(f, IOBufferSize)
-	}
-	bytesLeft := c.bytesLeft()
-	bytesToSend := c.chunkSize
-	if bytesLeft < int64(bytesToSend) {
-		bytesToSend = int(bytesLeft)
-	}
+
 	var data []byte
-	if c.contents == nil { // Data is being read from a file rather than static contents cache.
-		if c.offset == 0 && c.digest.Size <= int64(IOBufferSize) {
-			data = make([]byte, c.digest.Size)
-			c.contents = data // Cache the contents to avoid Reads on future Resets.
-		} else {
-			data = make([]byte, bytesToSend)
+	if c.contents != nil {
+		// As long as we have data in memory, it's much more efficient to return
+		// a view slice than to copy it around. Contents are immutable so it's okay
+		// to return the slice.
+		endRead := int(c.offset) + c.chunkSize
+		if endRead >= len(c.contents) {
+			endRead = len(c.contents)
+			c.reachedEOF = true
 		}
-		n, err := io.ReadFull(c.reader, data)
-		if err != nil {
-			return nil, err
+		data = c.contents[c.offset:endRead]
+	} else {
+		if !c.r.IsInitialized() {
+			err := c.r.Initialize()
+			if err != nil {
+				return nil, err
+			}
 		}
-		if n < bytesToSend {
-			return nil, fmt.Errorf("only read %d bytes from %s, expected %d", n, c.path, bytesToSend)
+
+		// We don't need to check the amount of bytes read, as ReadFull will yell if
+		// it's diff than len(data).
+		data = make([]byte, c.chunkSize)
+		n, err := io.ReadFull(c.r, data)
+		data = data[:n]
+		switch err {
+		case io.ErrUnexpectedEOF:
+			// Cache the contents to avoid further IO for small files.
+			if c.offset == 0 {
```

Comment:
Why not do this for both EOF cases?

Reply:
Hm, I did it like this because [ReadFull](https://golang.org/pkg/io/#ReadFull) only returns io.EOF if no bytes are read from `c.r`, which shouldn't realistically reach this point for `offset == 0` (it'll be caught at the Digest.Size == 0 check above, or it'll be some sort of error for the compressed bytes to be zero sized). 

In retrospect, the code is cleaner to read and not any less correct by grouping both behaviors. Done!